x/warp/contract.go

```diff
@@ -99,34 +112,168 @@ func getBlockchainID(accessibleState contract.AccessibleState, caller common.Add
+}
+
+// PackGetVerifiedWarpBlockHashOutput attempts to pack given [outputStruct] of type GetVerifiedWarpBlockHashOutput
+// to conform the ABI outputs.
+func PackGetVerifiedWarpBlockHashOutput(outputStruct GetVerifiedWarpBlockHashOutput) ([]byte, error) {
+	return WarpABI.PackOutput("getVerifiedWarpBlockHash",
+		outputStruct.WarpBlockHash,
+		outputStruct.Valid,
+	)
+}
+
+// UnpackGetVerifiedWarpBlockHashOutput attempts to unpack [output] as GetVerifiedWarpBlockHashOutput
+// assumes that [output] does not include selector (omits first 4 func signature bytes)
+func UnpackGetVerifiedWarpBlockHashOutput(output []byte) (GetVerifiedWarpBlockHashOutput, error) {
+	outputStruct := GetVerifiedWarpBlockHashOutput{}
+	err := WarpABI.UnpackIntoInterface(&outputStruct, "getVerifiedWarpBlockHash", output)
+
+	return outputStruct, err
+}
+
+func getVerifiedWarpBlockHash(accessibleState contract.AccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	remainingGas, err = contract.DeductGas(suppliedGas, GetVerifiedWarpMessageBaseCost)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+	predicateBytes, valid, err := unpackWarpMessage(accessibleState, input)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+	// If there is no such value or it failed verification, return invalid
+	if !valid {
+		packedOutput, err := PackGetVerifiedWarpBlockHashOutput(GetVerifiedWarpBlockHashOutput{
+			Valid: false,
+		})
+		if err != nil {
+			return nil, remainingGas, err
+		}
+		return packedOutput, remainingGas, nil
+	}
+
+	// Note: we charge for the size of the message during both predicate verification and each time the message is read during
+	// EVM execution because each execution incurs an additional read cost.
+	msgBytesGas, overflow := math.SafeMul(GasCostPerWarpMessageBytes, uint64(len(predicateBytes)))
+	if overflow {
+		return nil, remainingGas, vmerrs.ErrOutOfGas
+	}
+	if remainingGas, err = contract.DeductGas(remainingGas, msgBytesGas); err != nil {
+		return nil, 0, err
+	}
+	// Note: since the predicate is verified in advance of execution, the precompile should not
+	// hit an error during execution.
+	unpackedPredicateBytes, err := predicateutils.UnpackPredicate(predicateBytes)
+	if err != nil {
+		return nil, remainingGas, fmt.Errorf("%w: %s", errInvalidPredicateBytes, err)
+	}
+	warpMessage, err := warp.ParseMessage(unpackedPredicateBytes)
+	if err != nil {
+		return nil, remainingGas, fmt.Errorf("%w: %s", errInvalidWarpMsg, err)
+	}
+
+	blockHashPayload, err := warpPayload.ParseBlockHashPayload(warpMessage.UnsignedMessage.Payload)
+	if err != nil {
+		return nil, remainingGas, fmt.Errorf("%w: %s", errInvalidBlockHashPayload, err)
+	}
+	packedOutput, err := PackGetVerifiedWarpBlockHashOutput(GetVerifiedWarpBlockHashOutput{
+		WarpBlockHash: WarpBlockHash{
+			OriginChainID: common.Hash(warpMessage.SourceChainID),
```

Comment:
nit: could be more consistent to also name this `SourceChainID`

Reply:
Ya that's a good idea